pkg/build/api/types.go

```diff
@@ -20,6 +20,30 @@ const (
 	DefaultDockerLabelNamespace = "io.openshift."
 )
 
+const (
+	// SecretTypeSource contains data needed to authenticate against SCM servers.
+	//
+	// SecretTypeSource keys are devided into these option groups:
+	//   Basic authentication group. Requires at least one of fields:
+	//     - Secret.Data["username"] - username used to authenticate against SCM server
+	//     - Secret.Data["password"] - password or token needed to authenticate against SCM server
+	//     - Secret.Data["ca.crt"] - CA certificate needed to authenticate against SCM server
+	//   SSH authentication group. Required field:
+	//     - Secret.Data["ssh-privatekey"] - private SSH key needed to authenticate against SCM server
+	SecretTypeSource kapi.SecretType = "Source"
```

Comment:
On Sep 9, 2015, at 8:57 AM, Cesar Wong <notifications@github.com> wrote:

In pkg/build/api/types.go
<https://github.com/openshift/origin/pull/4433#discussion_r39038019>:

> @@ -20,6 +20,30 @@ const (
>  	DefaultDockerLabelNamespace = "io.openshift."
>  )
>
> +const (
> +	// SecretTypeSource contains data needed to authenticate against SCM servers.
> +	//
> +	// SecretTypeSource keys are devided into these option groups:
> +	//   Basic authentication group. Requires at least one of fields:
> +	//     - Secret.Data["username"] - username used to authenticate against SCM server
> +	//     - Secret.Data["password"] - password or token needed to authenticate against SCM server
> +	//     - Secret.Data["ca.crt"] - CA certificate needed to authenticate against SCM server
> +	//   SSH authentication group. Required field:
> +	//     - Secret.Data["ssh-privatekey"] - private SSH key needed to authenticate against SCM server
> +	SecretTypeSource kapi.SecretType = "Source"

Got it, so
1) Because uname/password and ssh-privatekey are mutually exclusive, we
should give them different types
2) The types have to be defined in kube.

To move this forward on our side, can we for now not declare a secret type
(use opaque)?



Yes, in fact, it would be generally useful for you to perform duck typing
in the absence of a secret type you recognize.  If someone gives us a
secret with a username and a password and we reject it because it's type "
mycompany.com/username" we're being jerks.


—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/4433/files#r39038019>.